src/main/perf/LineFileDocs.java

```diff
@@ -745,28 +751,71 @@ public Document nextDoc(DocState doc) throws IOException {
     }
   }
 
-  private static class LineFileDoc {
-    final FloatBuffer vector;
-    final String stringText;
-    final ByteBuffer byteText;
+  private static abstract class LineFileDoc {
+    private final FloatBuffer vector;
 
-    LineFileDoc(String text, float[] vector) {
-      stringText = text;
-      byteText = null;
+    LineFileDoc(float[] vector) {
       if (vector == null) {
         this.vector = null;
       } else {
         this.vector = FloatBuffer.wrap(vector);
       }
     }
 
-    LineFileDoc(ByteBuffer bytes, float[] vector) {
-      stringText = null;
-      byteText = bytes;
-      if (vector == null) {
-        this.vector = null;
-      } else {
-        this.vector = FloatBuffer.wrap(vector);
+    abstract int getId();
+
+    String getStringText() {
+      throw new UnsupportedOperationException();
+    }
+
+    ByteBuffer getByteText() {
+      throw new UnsupportedOperationException();
+    }
+
+    private static final class TextBased extends LineFileDoc {
+      final String stringText;
+
+      final int id; // will be the exact id if this is txt based LFD
+      TextBased(String text, float[] vector, int id) {
+        super(vector);
+        stringText = text;
+        this.id = id;
+      }
+
+      @Override
+      int getId() {
+        return id;
+      }
+
+      @Override
+      String getStringText() {
+        return stringText;
+      }
+    }
+
+    private static final class BinaryBased extends LineFileDoc {
+
+      final ByteBuffer byteText;
+      private int nextId; // will have multiple doc in a same LFD so we'll determine id using a base
+      private int docCount; // and a count
+      BinaryBased(ByteBuffer bytes, float[] vector, int idBase, int docCount) {
+        super(vector);
+        byteText = bytes;
+        this.nextId = idBase;
+        this.docCount = docCount;
+      }
+
+      @Override
+      int getId() {
+        if (docCount-- == 0) {
+          throw new IllegalStateException("Calling getId more than docCount");
+        }
+        return nextId++;
+      }
+
+      @Override
+      ByteBuffer getByteText() {
```

Comment:
I'm a bit confused by this class: `getId` impl seems to imply it iterates over the N docs in one block.  But, the vector and `byteText` are a single value as if for a single doc?

Reply:
Yeah ideally it would decode and generate the doc information one by one. But there're some differences in the way we decode and set the datetime it seems, so I only wrapped a very thin layer over the previous usage and not refactoring very deep.